### PR TITLE
[FrameworkBundle][HttpClient] Refactor http_client decoration strategy

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | https://github.com/symfony/symfony/pull/49302#discussion_r1115635023
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/HttpClient/DependencyInjection/HttpClientPass.php [hunk 1]
```diff
@@ -26,6 +26,11 @@ public function process(ContainerBuilder $container): void
         }
 
         foreach ($container->findTaggedServiceIds('http_client.client') as $id => $tags) {
+            // Compatibility with FrameworkBundle
+            if ('.http_client' === $id) {
```
Comment (fancyweb) [inline src/Symfony/Component/HttpClient/DependencyInjection/HttpClientPass.php]: @stof I hope you'll have a cleverer solution than this hack :grin:

**Comment by nicolas-grekas** [comment]
Rebase unlocked after #49302

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -2394,17 +2387,17 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
...
             } else {
                 $container->register($name, ScopingHttpClient::class)
-                    ->setArguments([new Reference($httpClientId), [$scope => $scopeConfig], $scope])
+                    ->setArguments([new Reference('http_client.transport'), [$scope => $scopeConfig], $scope])
                     ->addTag('http_client.client')
                 ;
             }
 
-            if ($this->readConfigEnabled('http_client.scoped_clients.'.$name.'retry_failed', $container, $retryOptions)) {
+            if ($this->readConfigEnabled('http_client.scoped_clients.'.$name.'.retry_failed', $container, $retryOptions)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:2400]: ~needs a fix on 6.2~

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -2413,7 +2406,7 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
                     ->register($name.'.uri_template', UriTemplateHttpClient::class)
                     ->setDecoratedService($name, null, 7) // Between TraceableHttpClient (5) and RetryableHttpClient (10)
                     ->setArguments([
-                        new Reference('.inner'),
+                        new Reference($name.'.uri_template.inner'),
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:2409]: let's be consistent with the rest of the file

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/http_client.php [hunk 4]
```diff
@@ -32,6 +32,10 @@
             ->call('setLogger', [service('logger')->ignoreOnInvalid()])
             ->tag('monolog.logger', ['channel' => 'http_client'])
             ->tag('kernel.reset', ['method' => 'reset', 'on_invalid' => 'ignore'])
+
+        ->set('http_client', HttpClientInterface::class)
+            ->factory('current')
+            ->args([[service('http_client.transport')]])
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/http_client.php:38]: Here is the master trick of this PR :)
This creates what I'd call a "tagged alias".

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @fancyweb.

**Pull request merged**